src/Poseidon/Janno.hs

```diff
@@ -394,15 +395,41 @@ createMinimalSample (EigenstratIndEntry id_ sex pop) =
 
 -- Janno file writing
 
+-- | A helper functions to replace empty bytestrings values in janno files with explicit "n/a"
+explicitNA :: Bch.ByteString -> Bch.ByteString
+explicitNA = replaceInJannoBytestring Bch.empty "n/a"
+
+replaceInJannoBytestring :: Bch.ByteString -> Bch.ByteString -> Bch.ByteString -> Bch.ByteString
+replaceInJannoBytestring from to tsv =
+    let tsvRows = Bch.lines tsv
+        tsvCells = map (Bch.splitWith (=='\t')) tsvRows
+        tsvCellsUpdated = map (map (\y -> if y == from || y == Bch.append from "\r" then to else y)) tsvCells
+        tsvRowsUpdated = map (Bch.intercalate (Bch.pack "\t")) tsvCellsUpdated
+   in Bch.unlines tsvRowsUpdated
+
+makeHeaderWithAdditionalColumns :: [JannoRow] -> Csv.Header
+makeHeaderWithAdditionalColumns rows =
+    V.fromList $ jannoHeader ++ sort (HM.keys (HM.unions (map (getCsvNR . jAdditionalColumns) rows)))
+
 writeJannoFile :: FilePath -> JannoRows -> IO ()
 writeJannoFile path (JannoRows rows) = do
-    let jannoAsBytestring = Csv.encodeByNameWith encodingOptions makeHeaderWithAdditionalColumns rows
-    let jannoAsBytestringwithNA = explicitNA jannoAsBytestring
+    let jannoAsBytestring = Csv.encodeByNameWith encodingOptions (makeHeaderWithAdditionalColumns rows) rows
+        jannoAsBytestringwithNA = explicitNA jannoAsBytestring
     Bch.writeFile path jannoAsBytestringwithNA
-    where
-        makeHeaderWithAdditionalColumns :: Csv.Header
-        makeHeaderWithAdditionalColumns =
-            V.fromList $ jannoHeader ++ sort (HM.keys (HM.unions (map (getCsvNR . jAdditionalColumns) rows)))
+
+writeJannoFileWithoutEmptyCols :: FilePath -> JannoRows -> IO ()
+writeJannoFileWithoutEmptyCols path (JannoRows rows) = do
+    let jannoAsBytestring = Csv.encodeByNameWith encodingOptions (makeHeaderWithAdditionalColumns rows) rows
+        jannoAsBytestringwithNA = explicitNA jannoAsBytestring
+    case Csv.decodeWith decodingOptions Csv.NoHeader jannoAsBytestringwithNA :: Either String (V.Vector (V.Vector Bch.ByteString)) of
+        Left _  -> error "internal error, please report"
+        Right x -> do
+            let janno = V.toList $ V.map V.toList x
+                jannoTransposed = transpose janno
```

Comment:
Clever! I didn't know Cassava can simply parse into a vector of vectors. That's great. I think in that case we could just improve `replaceInJannoBytestring` above using that same parsing code, to catch empty strings? Or build it into here.

Reply:
Right - that would be one possible solution.